xprof/tally.hpp

```diff
@@ -1,31 +1,820 @@
+*/
+template <typename TC, class... T, typename = std::enable_if_t<std::is_base_of_v<TallyCoreBase, TC>>>
+auto aggregate_nested(std::unordered_map<std::tuple<T...>, TC> &m) {
+
+  // New type for a tuple without the last element.
+  // Reference: https://stackoverflow.com/a/42043006/7674852
+  typedef decltype(make_tuple_cuted(std::declval<std::tuple<T...>>())) Minusone;
+
+  // Umap for the aggregated data
+  std::unordered_map<Minusone, std::unordered_map<thapi_function_name, TC>> aggregated;
+
+  for (auto &[key, val] : m) {
+    auto head = make_tuple_cuted(key);
+    aggregated[head][std::get<sizeof...(T) - 1>(key)] += val;
+  }
+  return aggregated;
+}
+
+//! Aggregate data by thapi_function_name.
+/*! This is used eventually to print in compact mode, i.e., statistics per thapi_function_name.
+\param m (std::unordered_map).
+\return Returns unordered map with data aggregated by thapi_function_name. 
+EXAMPLE:
+    input   umap{ 
+                  ("iris01",232,789,"getDeviceInfo") : CoreTime,
+                  ("iris02",123,890,"getDeviceInfo") : CoreTime 
+            }
+    output  umap{ "getDeviceInfio" : CoreTime } 
+*/
+template <typename TC, class... T, typename = std::enable_if_t<std::is_base_of_v<TallyCoreBase, TC>>>
+auto aggregate_by_name(std::unordered_map<std::tuple<T...>, TC> &m) {
+  std::unordered_map<thapi_function_name, TC> aggregated{};
+  
+  for (auto const &[key, val] : m)
+    // use the thapi_function_name as the key.
+    // thapi_function_name is the last element in the tuple "key".
+    aggregated[std::get<sizeof...(T) - 1>(key)] += val;
+  
+  return aggregated;
+}
+
+//! Add the elements of the tuple "t" in the set elements "s".
+/*! 
+\param s tuple of sets.
+\param t hpt_function_name_t (tuple)
+\param index sequence
+
+EXAMPLE:
+  input  
+    tuple{ set {}, set{}, set{}, set{} }, tuple{"iris01",232,789,"getDeviceInfo"}, std::index_sequence<0,1,2,3>
+  output (update first param by reference) 
+    tuple{ set {"iris01"}, set{232}, set{789}, set{"getDeviceInfo"} }
+
+  The index sequence used as argument "std::index_sequence<0,1,2,3>" enables the following unfolding:
+
+  (
+    std::get<0>(s).insert(std::get<0>(t)),
+    std::get<1>(s).insert(std::get<1>(t)),
+    std::get<2>(s).insert(std::get<2>(t)),
+    std::get<3>(s).insert(std::get<3>(t))
+  );
+
+  which adds every element of the tuple to its corresponding set, this updating the tuple of sets params
+  provided by reference.
+
+*/
+template <class... T, class... T2, size_t... I>
+void add_to_set(std::tuple<T...> &s, std::tuple<T2...> t, std::index_sequence<I...>) {
+  (std::get<I>(s).insert(std::get<I>(t)), ...);
+}
+
+// NOTE: It seems this function is not doing anything, so lets see if 
+// it breacks something.
+// template <class... T, size_t... I>
+// void remove_neutral(std::tuple<std::set<T>...> &s, std::index_sequence<I...>) {
+//   (std::get<I>(s).erase (T{}), ...);
+// }
```

Comment:
What is the pointer of this comment?

Reply:
No point, some unused function i wasn't sure to delete. Now it is gone !!!